lib/veri.ml

```diff
@@ -0,0 +1,237 @@
+open Core_kernel.Std
+open Bap.Std
+open Bap_traces.Std
+open Trace
+
+module SM = Monad.State
+open SM
+
+type point = {
+  code : Chunk.t;
+  side : event list
+}
+
+type 'a u = 'a Bil.Result.u
+type 'a r = 'a Bil.Result.r
+type 'a e = (event option, 'a) Monad.State.t
+
+let value = Bil.Result.value
+let make_point code = {code; side = [];}
+let add_side p ev = {p with side = ev :: p.side}
+let do_nothing () = SM.return ()
+
+let create_move_event tag cell' data' =  
+  Value.create tag Move.({cell = cell'; data = data';})   
+
+let find tag evs cond =
+  let open Option in
+  List.find evs ~f:(fun ev -> match Value.get tag ev with
+      | None -> false
+      | Some mv -> cond mv) >>| fun ev -> Value.get_exn tag ev
+
+let create_mem_store = create_move_event Event.memory_store
+let create_mem_load  = create_move_event Event.memory_load
+let create_reg_read  = create_move_event Event.register_read
+let create_reg_write = create_move_event Event.register_write
+let find_reg_read = find Event.register_read
+let find_mem_load = find Event.memory_load
+
+let is_imm_var var = 
+  let open Type in
+  match Var.typ var with
+  | Imm _ -> true
+  | _ -> false
+
+let imm_result v = 
+  let open Bil in
+  match v with
+  | Imm x -> Some x
+  | _ -> None
+  
+module Set = Value.Set
+
+class context report trace = object(self:'s)
+  inherit Veri_traci.context trace as super
+  val report : Veri_report.t = report
+  val events : Set.t = Set.empty
+  val selves = None
+  val current = `Fst
+  val descr : string option = None
+
+  method! update var res = 
+    match selves with
+    | None -> super#update var res
+    | Some (main, shadow) ->
+      match current with
+      | `Fst -> {<selves = Some (main#update var res, shadow) >}
+      | `Snd -> {<selves = Some (main, shadow#update var res) >}
+
+  method! lookup var = 
+    match selves with
+    | None -> super#lookup var 
+    | Some (main, shadow) ->
+      match current with
+      | `Fst -> main#lookup var
+      | `Snd -> shadow#lookup var
+
+  method split = 
+    let self = {< events = Set.empty; descr = None >} in
+    {< selves = Some (self, self); current = `Fst >}
+
+  method register_event ev = match selves with
+    | None -> {< events = Set.add events ev >}
+    | Some (main, shadow) ->
+      match current with
+      | `Fst -> {<selves = Some (main#register_event ev, shadow) >}
+      | `Snd -> {<selves = Some (main, shadow#register_event ev) >}
+
+  method real_events = match selves with
+    | None -> []
+    | Some (c,_) -> c#events
+
+  method merge: 's = 
+    let is_exists_event ev evs = 
+      List.exists ~f:(fun ev' -> ev = ev') evs in   
+    let events = self#main_exn#events in
+    let events' = self#shadow_exn#events in
+    let wrong = List.fold_left ~init:0 
```

Comment:
You should move this logic away from the context. The context by itself should be closed for modification, and new policies can be added by parametrizing the report type. Also, this logic doesn't look correct for me. It looks like some kind of simplification. 

The context, should just record mismatching events, i.e., the diff between to sets of side-effects in the current feram. The first approximation would be store something like 
```
`Left of event set |  `Right of event set | `Both (event * event) set
```
Where both contains a set of events that occurred on both phases, but are not equal.